src/core_plugins/tag_cloud/public/tag_cloud.js

```diff
@@ -0,0 +1,337 @@
+import d3 from 'd3';
+import d3TagCloud from 'd3-cloud';
+import vislibComponentsSeedColorsProvider from 'ui/vislib/components/color/seed_colors';
+import {EventEmitter} from 'events';
+
+
+const ORIENTATIONS = {
+  single: () => 0,
+  rightAngled: (tag) => {
+    return hashCode(tag.text) % 2 * 90;
+  },
+  multi: (tag) => {
+    return (~~(hashCode(tag.text) % 12) * 15) - 90;
+  }
+};
+const D3_SCALING_FUNCTIONS = {
+  linear: d3.scale.linear(),
+  log: d3.scale.log(),
+  sqrt: d3.scale.sqrt()
+};
+
+
+class TagCloud extends EventEmitter {
+
+  constructor(domNode) {
+
+    super();
+
+    this._element = domNode;
+    this._d3SvgContainer = d3.select(this._element).append('svg');
+    this._svgGroup = this._d3SvgContainer.append('g');
+    this._size = [1, 1];
+    this.resize();
+
+    this._fontFamily = 'Impact';
+    this._fontStyle = 'normal';
+    this._fontWeight = 'normal';
+    this._orientations = 'single';
+    this._minFontSize = 10;
+    this._maxFontSize = 36;
+    this._textScale = 'linear';
+    this._spiral = 'archimedean';//layout shape
+    this._timeInterval = 1000;//time allowed for layout algorithm
+    this._padding = 5;
+
+
+  }
+
+
+  setOptions(options) {
+    if (JSON.stringify(options) === this._optionsAsString) {
+      return;
+    }
+
+    this._optionsAsString = JSON.stringify(options);
+    this._orientations = options.orientations;
+    this._minFontSize = Math.min(options.minFontSize, options.maxFontSize);
+    this._maxFontSize = Math.max(options.minFontSize, options.maxFontSize);
+    this._textScale = options.textScale;
+    this._invalidate(false);
+  }
+
+
+  resize() {
+
+    const newWidth = this._element.offsetWidth;
+    const newHeight = this._element.offsetHeight;
+
+    if (newWidth < 1 || newHeight < 1) {
+      return;
+    }
+
+    if (newWidth === this._size[0] && newHeight === this._size[1]) {
+      return;
+    }
+
+    const didOverflow = this._size[0] < this._cloudWidth || this._size[1] < this._cloudHeight;
+    const willNotOverflow = this._cloudWidth <= newWidth && this._cloudHeight <= newHeight;
+
+    this._size[0] = newWidth;
+    this._size[1] = newHeight;
+
+    if (!didOverflow && willNotOverflow && this._allInViewBox) {
+      this._invalidate(true);
+    } else {
+      this._invalidate(false);
+    }
+
+  }
+
+  setData(data) {
+    this._words = data.map(toWordTag);
+    this._makeTextSizeMapper();
+    this._invalidate(false);
+  }
+
+
+  destroy() {
+    clearTimeout(this._timeoutHandle);
+    this._element.innerHTML = '';
+  }
+
+  getStatus() {
+    return this._allInViewBox ? TagCloud.STATUS.COMPLETE : TagCloud.STATUS.INCOMPLETE;
+  }
+
+  async whenRendered() {
```

Comment:
as TagCloud already extends EventEmitter and you use tagCloud.on('select') to handle clicking/selecting a tag, what do you think about using the same functionality to notify about renderComplete as well ? I see a benefit of using a promise in case when you are not interested in future re-renderings, but this is a bit hard for me to read. Maybe changing the name of whenRendered to something else would help (don't have any good suggestions .. getRenderCompletePromise might be a bit long :)  )